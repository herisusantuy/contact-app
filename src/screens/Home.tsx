import {
  StyleSheet,
  FlatList,
  StatusBar,
  SafeAreaView,
  ActivityIndicator,
} from "react-native";
import React, { useEffect, useState } from "react";
import FloatingButton from "../components/floating-button";
import { useAppDispatch, useAppSelector } from "../redux/store/hooks";
import CardContact from "../components/card-contact";
import { getContactsAction } from "../redux/action/contactAction";
import FormModal from "../components/form-modal";
import { IContact, TContactBody } from "../redux/types";

export default function Home() {
  const { contacts, loading, error } = useAppSelector((state) => state.contact);
  const dispatch = useAppDispatch();

  const [isVisible, setIsVisible] = useState<boolean>(false);
  const [isEdit, setIsEdit] = useState<boolean>(false);
  const [selectedContact, setSelectedContact] = useState<IContact>();

  useEffect(() => {
    getContacts();
  }, []);

  const getContacts = async () => {
    const res = await dispatch(getContactsAction());
  };

  const handleUpdateContact = (contact: IContact) => {
    setSelectedContact(contact);
    setIsEdit(true);
    setIsVisible(true);
  };

  return (
    <SafeAreaView style={styles.container}>
      <FormModal
        isVisible={isVisible}
        onClose={() => setIsVisible(false)}
        isEdit={isEdit}
        contact={selectedContact}
      />
      {loading ? (
        <ActivityIndicator style={styles.spinner} />
      ) : (
        <FlatList
          data={contacts}
          renderItem={({ item }) => (
            <CardContact
              contact={item}
              onUpdateContact={() => handleUpdateContact(item)}
            />
          )}
          keyExtractor={(item) => item.id}
        />
      )}

      <FloatingButton onPress={() => setIsVisible(true)} />
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginTop: StatusBar.currentHeight || 0,
    backgroundColor: "white",
  },
  spinner: {
    marginTop: 20,
  },
});
