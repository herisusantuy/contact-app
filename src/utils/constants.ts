export const testID = {
  swipeableWrapper: "card-contact-swipeable-wrapper",
  photoProfile: "card-contact-photo-profile",
  fullnameText: "card-contact-fullname-text",
  ageText: "card-contact-age-text",
  pressableView: "floating-button-pressable-view",
  plusIcon: "floating-button-plus-icon",
  textInput: "form-text-field",
  customModal: "custom-modal",
  cancelButton: "cancel-button",
  saveButton: "save-button",
};
