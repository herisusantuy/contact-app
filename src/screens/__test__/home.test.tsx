import { describe, expect, test } from "@jest/globals";
import renderer from "react-test-renderer";
import Home from "../Home";
import { renderWithProviders } from "../../utils/test-utils";

describe("Home", () => {
  it("Home renders correctly", () => {
    const tree = renderWithProviders(<Home />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
