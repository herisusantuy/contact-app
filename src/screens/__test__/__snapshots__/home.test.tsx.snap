// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Home renders correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "white",
      "flex": 1,
      "marginTop": 0,
    }
  }
>
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    supportedOrientations={
      [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
    transparent={true}
    visible={false}
  />
  <ActivityIndicator
    style={
      {
        "marginTop": 20,
      }
    }
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "green",
        "borderColor": "green",
        "borderRadius": 100,
        "borderWidth": 1,
        "bottom": 100,
        "height": 70,
        "justifyContent": "center",
        "opacity": 1,
        "position": "absolute",
        "right": 30,
        "width": 70,
      }
    }
    testID="floating-button-pressable-view"
  >
    <Text />
  </View>
</RCTSafeAreaView>
`;
